File: packages/dart/test/curp_validator_test.dart
import 'package:test/test.dart';
import 'package:catalogmx/catalogmx.dart';

void main() {
  group('CURP Validator', () {
    test('validates valid CURPs', () {
      expect(validateCURP('OEAF771012HMCRGR09'), isTrue);
      expect(validateCURP('PEGJ900515HJCRRN05'), isTrue);
      expect(validateCURP('MEHL850101MMCRRS08'), isTrue);
    });

    test('rejects invalid CURPs', () {
      expect(validateCURP(''), isFalse);
      expect(validateCURP('INVALID'), isFalse);
      expect(validateCURP('OEAF771012HMCRGR0'), isFalse); // Too short
      expect(validateCURP('OEAF771012HMCRGR099'), isFalse); // Too long
    });

    test('rejects CURP with invalid structure', () {
      expect(validateCURP('1EAF771012HMCRGR09'), isFalse); // First char must be letter
      expect(validateCURP('OEAF991332HMCRGR09'), isFalse); // Invalid date
      expect(validateCURP('OEAF771012XMCRGR09'), isFalse); // Invalid gender
    });

    test('validates check digit', () {
      final validator = CURPValidator('OEAF771012HMCRGR09');
      expect(validator.validateCheckDigit(), isTrue);
    });

    test('extracts birth date', () {
      final validator = CURPValidator('OEAF771012HMCRGR09');
      final date = validator.getBirthDate();

      expect(date, isNotNull);
      expect(date!.year, equals(1977));
      expect(date.month, equals(10));
      expect(date.day, equals(12));
    });

    test('extracts gender', () {
      final validator1 = CURPValidator('OEAF771012HMCRGR09');
      expect(validator1.getGender(), equals('Hombre'));

      final validator2 = CURPValidator('MEHL850101MMCRRS08');
      expect(validator2.getGender(), equals('Mujer'));
    });

    test('extracts state code', () {
      final validator = CURPValidator('OEAF771012HMCRGR09');
      expect(validator.getStateCode(), equals('MC'));
    });
  });

  group('CURP Generator', () {
    test('generates valid CURP', () {
      final curp = generateCURP(
        nombre: 'Juan',
        apellidoPaterno: 'García',
        apellidoMaterno: 'López',
        fechaNacimiento: DateTime(1990, 5, 15),
        sexo: 'H',
        estado: 'Jalisco',
      );

      expect(curp.length, equals(18));
      expect(validateCURP(curp), isTrue);
    });

    test('generates CURP with accents', () {
      final curp = generateCURP(
        nombre: 'José María',
        apellidoPaterno: 'Pérez',
        apellidoMaterno: 'García',
        fechaNacimiento: DateTime(1990, 5, 15),
        sexo: 'H',
        estado: 'México',
      );

      expect(validateCURP(curp), isTrue);
    });

    test('generates CURP without materno', () {
      final curp = generateCURP(
        nombre: 'Juan',
        apellidoPaterno: 'García',
        apellidoMaterno: null,
        fechaNacimiento: DateTime(1990, 5, 15),
        sexo: 'H',
        estado: 'Jalisco',
      );

      expect(validateCURP(curp), isTrue);
      expect(curp[2], equals('X')); // Should have X for missing materno
    });

    test('generates CURP for female', () {
      final curp = generateCURP(
        nombre: 'María',
        apellidoPaterno: 'Hernández',
        apellidoMaterno: 'López',
        fechaNacimiento: DateTime(1985, 1, 1),
        sexo: 'M',
        estado: 'México',
      );

      expect(validateCURP(curp), isTrue);
      expect(curp[10], equals('M'));
    });

    test('handles state codes', () {
      final curp1 = generateCURP(
        nombre: 'Juan',
        apellidoPaterno: 'García',
        apellidoMaterno: 'López',
        fechaNacimiento: DateTime(1990, 5, 15),
        sexo: 'H',
        estado: 'Ciudad de México',
      );
      expect(curp1.substring(11, 13), equals('DF'));

      final curp2 = generateCURP(
        nombre: 'Juan',
        apellidoPaterno: 'García',
        apellidoMaterno: 'López',
        fechaNacimiento: DateTime(1990, 5, 15),
        sexo: 'H',
        estado: 'Jalisco',
      );
      expect(curp2.substring(11, 13), equals('JC'));
    });

    test('handles born abroad', () {
      final curp = generateCURP(
        nombre: 'Juan',
        apellidoPaterno: 'García',
        apellidoMaterno: 'López',
        fechaNacimiento: DateTime(1990, 5, 15),
        sexo: 'H',
        estado: null,
      );

      expect(curp.substring(11, 13), equals('NE'));
    });

    test('calculates check digit correctly', () {
      final checkDigit = CURPGenerator.calculateCheckDigit('OEAF771012HMCRGR0');
      expect(checkDigit, equals('9'));
    });
  });
}

File: packages/kotlin/src/test/kotlin/com/openbancor/catalogmx/validators/CurpValidatorTest.kt
package com.openbancor.catalogmx.validators

import com.openbancor.catalogmx.validateCURP
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test
import java.time.LocalDate

class CurpValidatorTest {

    @Test
    fun `should validate correct CURP`() {
        // Example CURPs with valid structure
        assertTrue(validateCURP("OEAF771012HMCRGR09"))
    }

    @Test
    fun `should reject CURP with invalid length`() {
        assertFalse(validateCURP("OEAF771012HMCRGR0"))  // 17 chars
        assertFalse(validateCURP("OEAF771012HMCRGR099")) // 19 chars
    }

    @Test
    fun `should reject CURP with invalid structure`() {
        assertFalse(validateCURP("1234567890123456789")) // All numbers
        assertFalse(validateCURP("ABCDEFGHIJKLMNOPQR")) // All letters
    }

    @Test
    fun `should extract gender from CURP`() {
        val validatorH = CurpValidator("OEAF771012HMCRGR09")
        assertEquals("Hombre", validatorH.getGender())

        val validatorM = CurpValidator("OEAF771012MMCRGR09")
        assertEquals("Mujer", validatorM.getGender())
    }

    @Test
    fun `should extract state code from CURP`() {
        val validator = CurpValidator("OEAF771012HMCRGR09")
        assertEquals("MC", validator.getStateCode())
    }

    @Test
    fun `should extract birth date from CURP`() {
        val validator = CurpValidator("OEAF771012HMCRGR09")
        val birthDate = validator.getBirthDate()
        assertNotNull(birthDate)
        assertEquals(1977, birthDate!!.year)
        assertEquals(10, birthDate.monthValue)
        assertEquals(12, birthDate.dayOfMonth)
    }

    @Test
    fun `should calculate check digit correctly`() {
        // The check digit calculation uses a dictionary of 37 chars
        val checkDigit = CurpGenerator.calculateCheckDigit("OEAF771012HMCRGR0")
        assertEquals('8', checkDigit)
    }

    @Test
    fun `should generate valid CURP`() {
        val curp = CurpGenerator(
            nombre = "JUAN",
            apellidoPaterno = "GARCIA",
            apellidoMaterno = "LOPEZ",
            fechaNacimiento = LocalDate.of(1990, 5, 15),
            sexo = "H",
            estado = "JALISCO"
        ).generate()

        assertEquals(18, curp.length)
        assertTrue(validateCURP(curp))

        // Check structure
        assertEquals("GALJ", curp.substring(0, 4)) // Name letters
        assertEquals("900515", curp.substring(4, 10)) // Birth date
        assertEquals("H", curp.substring(10, 11)) // Gender
        assertEquals("JC", curp.substring(11, 13)) // State code
    }

    @Test
    fun `should handle cacophonic names`() {
        val curp = CurpGenerator(
            nombre = "ANA",
            apellidoPaterno = "CACA", // Would be cacophonic but result is CALA, not CACA
            apellidoMaterno = "LOPEZ",
            fechaNacimiento = LocalDate.of(1990, 5, 15),
            sexo = "M",
            estado = "DF"
        ).generate()

        // Result is CALA (C + A vowel + L from Lopez + A from Ana)
        // CALA is not in cacophonic list, so it stays unchanged
        assertEquals("CALA", curp.substring(0, 4))
    }

    @Test
    fun `should handle names with MARIA or JOSE prefix`() {
        val curp = CurpGenerator(
            nombre = "MARIA FERNANDA",
            apellidoPaterno = "LOPEZ",
            apellidoMaterno = "HERNANDEZ",
            fechaNacimiento = LocalDate.of(1992, 1, 3),
            sexo = "M",
            estado = "DF"
        ).generate()

        // Should use FERNANDA, not MARIA
        assertEquals('F', curp[3])
    }

    @Test
    fun `should handle missing materno surname`() {
        val curp = CurpGenerator(
            nombre = "CARLOS",
            apellidoPaterno = "PEREZ",
            apellidoMaterno = null,
            fechaNacimiento = LocalDate.of(1980, 7, 9),
            sexo = "H",
            estado = "DF"
        ).generate()

        assertEquals(18, curp.length)
        assertTrue(validateCURP(curp))
        // Third position should be X when no materno
        assertEquals('X', curp[2])
    }

    @Test
    fun `state codes should be mapped correctly`() {
        val stateTests = mapOf(
            "AGUASCALIENTES" to "AS",
            "BAJA CALIFORNIA" to "BC",
            "CIUDAD DE MEXICO" to "DF",
            "CDMX" to "DF",
            "JALISCO" to "JC",
            "NUEVO LEON" to "NL",
            "EXTRANJERO" to "NE"
        )

        stateTests.forEach { (stateName, expectedCode) ->
            assertTrue(
                CurpValidator.stateCodes.containsKey(stateName),
                "State $stateName should be in stateCodes map"
            )
            assertEquals(expectedCode, CurpValidator.stateCodes[stateName])
        }
    }
}
